src/azure-cli-core/azure/cli/core/util.py

```diff
@@ -584,6 +584,10 @@ def send_raw_request(cli_ctx, method, uri, headers=None, uri_parameters=None,  #
     if '{subscriptionId}' in uri:
         uri = uri.replace('{subscriptionId}', profile.get_subscription_id())
 
+    # Dump custom request headers before adding Authorization header
+    logger.info("Request headers:")
```

Comment:
please have output be valid JSON, so I can easily programmatically parse the entire output

Reply:
I agree. The problem is there are two parts - request and response headers. We want to display request headers before sending the request, so they have to be torn apart. 

Also, `stderr` may contain other trash like `command ran in 4.608 seconds.`, so it is not a good practice to extract info from `stderr`. 

For the `Authorization` header, the current behavior of msrest with `--debug` is to totally remove it. We previously also removed similar information with https://github.com/Azure/azure-cli/pull/10473 due to security concerns. So currently let's just hide it until we get approval from PM @achandmsft .

What I am thinking is to add a param `--include-headers`, so that result in `stdout` will become something like:

```jsonc
{
    "request": {
        "headers": {
            "User-Agent": "AZURECLI/2.0.81",
            "x-ms-client-request-id": "b43346df-c51f-475a-bf7d-eec64d394f15",
            "CommandName": "rest",
            "ParameterSetName": "-u --verbose"
        }
    },
    "response": {
        "headers": {
            "Cache-Control": "no-cache",
            "Pragma": "no-cache",
            "Content-Type": "application/json; charset=utf-8",
            "Content-Encoding": "gzip",
            "Expires": "-1",
            "Vary": "Accept-Encoding",
            "x-ms-ratelimit-remaining-subscription-reads": "11999",
            "x-ms-request-id": "05761bc6-1b26-44be-a676-a6999e950a8a",
            "x-ms-correlation-request-id": "05761bc6-1b26-44be-a676-a6999e950a8a",
            "x-ms-routing-request-id": "EASTASIA:20200210T113239Z:05761bc6-1b26-44be-a676-a6999e950a8a",
            "Strict-Transport-Security": "max-age=31536000; includeSubDomains",
            "X-Content-Type-Options": "nosniff",
            "Date": "Mon, 10 Feb 2020 11:32:38 GMT",
            "Content-Length": "279"
        },
        "status": 200,
        "body": {
            "id": "/subscriptions/0b1f6471-1bf0-4dda-aec3-cb9272f09590/resourceGroups/xxx",
            "location": "southeastasia",
            "name": "xxx",
            "properties": {
                "provisioningState": "Succeeded"
            },
            "type": "Microsoft.Resources/resourceGroups"
        }
    }
}
```

In thus case, the output can be easily scripted from `stdout` instead of `stderr`. Any suggestion is welcome.

Reply:
I like stdout and full json, just make sure it is escaped properly, i.e. if body contains quotes

--include-headers should be something like the print flag in the httpie library that I referenced in the bug.  Sometimes you only want certain headers.

For auth, I would still include the header, so the user knows the header was there, but redact the actual token.

`Authorization: Bearer ey=oo97er...0789er`

Not the full token.

Reply:
Let's address the full json output (including headers) in another PR. I have updated this one to follow the format from msrest (see the updated description).

For `Authentication` header, I have included it in the log and trimmed it to only keep the first 15 characters. Does this suit the needs?

```
'Authorization': 'Bearer eyJ0eXAi*****'
```

Reply:
@jiasli any update on been able to print `--verbose` logs as JSON ? That will be useful to parse it :)

Reply:
@jiasli any update on the `--include-headers` proposal?